src/rules.jl

```diff
@@ -96,3 +96,110 @@ true
 See also: [`frule`](@ref), [`@scalar_rule`](@ref)
 """
 rrule(::Any, ::Vararg{Any}; kwargs...) = nothing
+
+
+#######################################################################
+# Infastructure to support generating overloads from rules.
+
+const NEW_RRULE_HOOKS = Function[]
+const NEW_FRULE_HOOKS = Function[]
+_hook_list(::typeof(rrule)) = NEW_RRULE_HOOKS
+_hook_list(::typeof(frule)) = NEW_FRULE_HOOKS
+
+"""
+    on_new_rule(sig->eval...), frule | rrule)
+"""
+function on_new_rule(hook_fun, rule_kind)
+    # get all the existing rules
+    ret = map(_rule_list(rule_kind)) do method
+        sig = _primal_sig(rule_kind, method)
+        _safe_hook_fun(hook_fun, sig)
+    end
+
+    # register hook for new rules
+    push!(_hook_list(rule_kind), hook_fun)
+    return ret
+end
+
+function __init__()
+    push!(Base.package_callbacks, pkgid -> refresh_rules())
+    push!(Base.include_callbacks, (mod, filename) -> refresh_rules())
+end
+
+
+"""
+    _rule_list(frule | rrule)
```

Comment:
very minor, but i wonder if `_methods` is a better name (since it wraps `methods` so closely)

Reply:
I don't think so,
i think that is a implementation detail.
IIRC an earlier version of this PR had this working entirely differently, and returning something different but the places it was uses remain the same